verification/src/http_server/handlers/verification/solidity/handlers.rs

```diff
@@ -0,0 +1,106 @@
+use crate::{
+    compiler::{CompilerVersion, Compilers, CompilersError, Fetcher},
+    solidity::{VerificationSuccess, Verifier, VerifierInitializationError},
+    VerificationResponse, VerificationResult,
+};
+use actix_web::error;
+use ethers_solc::{
+    artifacts::{BytecodeHash, SettingsMetadata},
+    CompilerInput,
+};
+use std::{
+    fmt::{Debug, Display},
+    str::FromStr,
+};
+use thiserror::Error;
+
+const BYTECODE_HASHES: [BytecodeHash; 3] =
+    [BytecodeHash::Ipfs, BytecodeHash::None, BytecodeHash::Bzzr1];
+
+pub struct CompileAndVerifyInput<'a> {
+    pub compiler_version: CompilerVersion,
+    pub compiler_input: CompilerInput,
+    pub creation_tx_input: &'a str,
+    pub deployed_bytecode: &'a str,
+}
+
+#[derive(Error, Debug)]
+enum CompileAndVerifyError {
+    #[error("{0:#}")]
+    VerifierInitialization(#[from] VerifierInitializationError),
+    #[error("{0:#}")]
+    Compilation(#[from] CompilersError),
+    #[error("No contract could be verified with provided data")]
+    NoMatchingContracts,
+}
+
+pub(crate) async fn compile_and_verify_handler<T: Fetcher>(
+    compilers: &Compilers<T>,
+    mut input: CompileAndVerifyInput<'_>,
+    bruteforce_bytecode_hashes: bool,
+) -> Result<VerificationResponse, actix_web::Error>
+where
+    <T as Fetcher>::Error: Debug + Display,
+{
+    let bruteforce_metadata = if !bruteforce_bytecode_hashes {
+        Vec::from([input.compiler_input.settings.metadata.clone()])
+    } else if pre_v0_6_0(&input.compiler_version) {
+        Vec::from([None])
+    } else {
+        Vec::from(BYTECODE_HASHES.map(|hash| Some(SettingsMetadata::from(hash))))
+    };
+
+    for metadata in bruteforce_metadata {
+        input.compiler_input.settings.metadata = metadata;
+        match compile_and_verify(compilers, &input).await {
+            Ok(verification_success) => {
+                let verification_result = VerificationResult::from((
+                    input.compiler_input,
+                    input.compiler_version,
+                    verification_success,
+                ));
+                return Ok(VerificationResponse::ok(verification_result));
+            }
+            Err(CompileAndVerifyError::VerifierInitialization(err)) => {
+                return Err(error::ErrorBadRequest(err))
+            }
+            Err(CompileAndVerifyError::Compilation(CompilersError::Fetch(err))) => {
```

Comment:
I don't like idea of `Err(...)` here. I think that user input errors should be returned as `VerificationResponse::err`.

`Err(...)` should be use in case of very strange error that shouldn't happen (like invalid bytecode, or invalid compiler versions)

Reply:
There are different kind of errors that may occur here. `CompilersError::Fetch` may be returned in two cases: 1) if user supplied invalid compiler version which is not in our list of all versions, and 2) if user provided valid version but downloading for some reason failed (e.g. link for download is invalid). In case (2) we certainly should return `InternalError`. In case (1) however (I believe that is the case you've meant), I believe we should return not 200 with `VerificationResponse::err` but  `400 BadRequest`. It is because we provide the user with a list of compilers we are ready to accept, and if they still provide us something invalid should be considered as their fault, I believe.